### ISSUE TITLE
[EKS] [request]: Allow specifying which admission controllers are enabled

### ISSUE THREAD

**Issue by hendrikhalkow** [aws/containers-roadmap#739]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
When I create an EKS cluster, I'd like to specify which of the _compiled-in_ admission controllers are enabled.

**Which service(s) is this request for?**
EKS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
According to the [documentation](https://docs.aws.amazon.com/eks/latest/userguide/platform-versions.html), there is an immutable list of admission controllers that are enabled by default for each EKS version. I'd like to use the _AlwaysPullImages_ admission controller without using [dynamic admission control](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/).

**Are you currently working around this issue?**
Since the _AlwaysPullImages_ is the only admission controller for which I'd like to have this, I use immutable image tags.

**Comment by brychcy** [comment]
It should be possible to enable the ExtendedResourceToleration admission controller, which is according to https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/#example-use-cases the recommended way to restrict GPU nodes to pods that need them.

**Comment by SQUIDwarrior** [comment]
Yes this is essential to allowing Spark users to schedule GPUs on EKS since Spark 2.4 does not provide a way to modify the Kubernetes operations manually.

**Comment by rajakshay** [comment]
> It should be possible to enable the ExtendedResourceToleration admission controller, which is according to https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/#example-use-cases the recommended way to restrict GPU nodes to pods that need them.

Not sure if you meant that it is already possible to enable `ExtendedResourceToleration` admission controller in EKS ?

**Comment by shinji62** [comment]
@mikestef9 anything we can do to bump this issue ?

**Comment by mikestef9** [comment]
We will be enabling the ExtendedResourceToleration admission controller with the upcoming 1.19 launch. Are there any other admission controllers you are interested in?

**Comment by Elgarni** [comment]
@mikestef9 I would also like to see `PodNodeSelector` enabled, to be able to select nodes on the namespace level,as described in [k8s docs here](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#configuration-annotation-format)

**Comment by evanstucker-hates-2fa** [comment]
I want the AlwaysPullImages controller so our multi-tenant dev clusters are more secure and so I don't have to add `imagePullPolicy: Always` to everything everywhere - it will just be `Always` by default. https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#alwayspullimages

**Comment by joshpaul-okta** [comment]
We are interested in the following two admission controllers as used together allow for explicit inherited and enforced placement of workloads at the namespace level.

https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#podnodeselector
https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#podtolerationrestriction

Adding the labels to the namespace the restricts all resources in the namespace to the target nodegroup and will cause them to be inherited so teams do not have to work about this.  With the additional use of taints on the nodegroups we get guarantees on workloads not being able to drift from the targeted capacity.

**Comment by ryanisnan** [comment]
Our use-case is similar to @joshpaul-okta.

We'd really like to be able to use the `PodTolerationRestriction` admission controller. We have a managed node-group which we'd like to reserve for a wide-number of community-provided helm chart resources. The helm charts that we're including can be really complex, and having to add tolerations manually is tedious, error prone, and not very resilient to change.

Being able to add tolerations to pods by namespace configuration would make this go from being a real nightmare of a problem, to something I never think about.

**Comment by evanstucker-hates-2fa** [comment]
`imagePullPolicy: Always` means that every time the kubelet launches a container, the kubelet queries the container image registry to resolve the name to an image digest. If the kubelet has a container image with that exact digest cached locally, the kubelet uses its cached image; otherwise, the kubelet pulls the image with the resolved digest, and uses that image to launch the container (https://kubernetes.io/docs/concepts/containers/images/#image-pull-policy).

This is desirable for security for a few reasons:

- "Relying on cached versions of a Docker image can become a security vulnerability. By default, an image will be pulled if it isn't already cached on the node attempting to run it. This can result in variations in images that are running per node, or potentially provide a way to gain access to an image without having direct access to the ImagePullSecret." (https://www.fairwinds.com/blog/kubernetes-how-to-ensure-imagepullpolicy-set-to-always)

- "Admin1 creates Pod1 that uses SuperSecretImage1 by specifying ImagePullSecret1 to gain access to the SecureContainerRegistry holding the image. Kubernetes pulls down and caches SuperSecretImage1 on the node and then creates Pod1 accordingly. Admin2, who does not have access to ImagePullSecret1 or SecureContainerRegistry, attempts to create Pod2 using SuperSecretImage1 and an imagePullPolicy of IfNotPresent. Because this image is cached on the node, this operation is successful." (https://trstringer.com/kubernetes-alwayspullimages/) The mitigation for this is to use an AdmissionController that forces ImagePullPolicy to Always on all workloads.

We've been waiting for this for almost 3 years now. I suppose it’s not as critical for me in particular now that we have separate clusters per team... And it’s definitely more profitable for AWS to push people towards a cluster-per-team paradigm, instead of making it easier to have efficient, secure multi-tenant clusters.

Are we ever going to see this feature added?

**Comment by daniel-laszlo** [comment]
Another usecase is enforcing Pod Security Standards compliance clusterwide: https://kubernetes.io/docs/tasks/configure-pod-container/enforce-standards-admission-controller/
